oscaralanpierce/skyrim_inventory_management_frontend
# Enable games to be created from the shopping list page

## Context

* [**Ensure that shopping list loading stops when there are no games**](https://trello.com/c/fgyaX7pS/284-ensure-that-shopping-list-loading-stops-when-there-are-no-games)
* [**Enable games to be created from the shopping/inventory list pages**](https://trello.com/c/fvsTSrEb/275-enable-games-to-be-created-from-the-shopping-inventory-list-pages)

Currently, if a user doesn't have games and they visit the shopping lists page, the shopping list pulse loader component will never go away. We want this component to go away when the games have finished loading and be replaced by a message prompting the user to create a game. Clicking a link in the prompt should cause a modal game create form to appear. When the form is successfully submitted, the newly-created game should become the active game and its shopping lists - which there will be none of - should be fetched from the API.

## Changes

* Replace `GameEditForm` component with more flexible `GameForm` component that can be used for either creation or editing
* Add logic to `ShoppingListGrouping` component to display a message that games are required to use the shopping lists page, and a link to show the create form
* Make page gutters the same size as header margin at small viewport widths
* Update tests and Storybook stories

## Required Tasks


## Manual Test Cases

You will need to check that functionality for creating a game from the shopping lists page works, as well as that the game edit functionality has not been broken in the process.

### Creating a Game from the Shopping Lists Page

For these tests, you will need to be logged in as a user with no games.

#### Happy Path

1. On the shopping lists page, see a message that you need a game to use the shopping lists feature and a link to create a game
2. Click the link
3. See that the modal form appears
4. Submit the form, with or without updating the fields
5. See that the name of your new game now appears as the active game in the `StyledSelect` dropdown
6. See the message "This game has no shopping lists."
7. Verify that the shopping list creation form is enabled
8. Check the API logs to verify that shopping lists were requested for the new game

#### Unprocessable Entity

These errors should never occur due to the impossibility of uniqueness errors and the fact that name format issues will be picked up by form validations before submitting the form.

1. Program the API to always return a 422 response from the `GamesController::CreateService#perform` method
2. Follow steps 1-4 under "Happy Path", above
3. See a flash error message with validation errors returned from the server
4. See that the form is not hidden or cleared

#### Internal Server Error

These errors are also never expected to occur.

1. Program the API to always return a 500 response from the `GamesController::CreateService#perform` method
2. Follow steps 1-4 under "Happy Path", above
3. See a flash error message informing the user something went wrong
4. See that the flash message doesn't include the exact error returned from the server
5. See that the form is not hidden or cleared

### Updating a Game on the Games Page

#### Happy Path

1. Once you have created a game, navigate to the games page
2. Click the edit icon next to your game's name
3. See that the edit form appears with the existing name and description (if any) autopopulated
4. Change the name or description of your game to a valid value
5. Submit the form
6. See that your game is updated on the list
7. See that the form and modal are hidden
8. See a flash success message that your game has been updated successfully

#### Unprocessable Entity

1. On the games page, create a second game
2. Click the edit icon next to one of your games
3. See that the edit form appears with the existing attributes of that game autopopulated
4. Change the name of the game to the name of your other game
5. Submit the form
6. See that the form is not hidden or cleared
7. See a flash error message indicating the name must be unique

#### Internal Server Error

1. Program the API to always return a 500 response from the `GamesController::UpdateService#perform` method
2. Follow steps 1-5 under "Happy Path", above
3. See a non-technical error message indicating something went wrong
4. See that the form is not hidden or cleared

## Screenshots and GIFs

### 320px

#### Form Hidden

<img width="320" alt="ShoppingListsEmpty-320" src="https://user-images.githubusercontent.com/5115928/234139114-9f70e65d-3c3a-4b85-a67f-d054b16c6a01.png">

#### Form Visible

<img width="320" alt="ShoppingListsForm-320" src="https://user-images.githubusercontent.com/5115928/234139149-cafe69ec-1cec-4d36-9822-4326d4376cb3.png">

### 481px

#### Form Hidden

<img width="481" alt="ShoppingListsEmpty-481" src="https://user-images.githubusercontent.com/5115928/234139169-68408876-e48f-452b-937b-0a2665a892c9.png">

#### Form Visible

<img width="481" alt="ShoppingListsForm-481" src="https://user-images.githubusercontent.com/5115928/234139192-86e50c41-0e1a-4c6e-a3fc-293292927e3e.png">

### 601px

#### Form Hidden

<img width="601" alt="ShoppingListsEmpty-601" src="https://user-images.githubusercontent.com/5115928/234139223-73c0f997-5c9c-44ac-8914-615c870a1926.png">

#### Form Visible

<img width="601" alt="ShoppingListsForm-601" src="https://user-images.githubusercontent.com/5115928/234139243-fcef83c3-4d9b-4d4a-8fe4-56b510e5fba8.png">

### 769px

#### Form Hidden

<img width="769" alt="ShoppingListsEmpty-769" src="https://user-images.githubusercontent.com/5115928/234139277-4c8050db-5371-486f-a701-5e74bbd1fe5e.png">

#### Form Visible

<img width="769" alt="ShoppingListsForm-769" src="https://user-images.githubusercontent.com/5115928/234139301-666325ce-3168-43c6-b0c6-75935fd84174.png">

### 1025px

#### Form Hidden

<img width="1025" alt="ShoppingListsEmpty-1025" src="https://user-images.githubusercontent.com/5115928/234139324-cad9eef6-9df1-41f4-9f67-8ec26d057ee5.png">

#### Form Visible

<img width="1025" alt="ShoppingListsForm-1025" src="https://user-images.githubusercontent.com/5115928/234139350-7f7412f1-9b65-4f5d-9acd-785f0113d59b.png">

### 1201px

#### Form Hidden

<img width="1201" alt="ShoppingListsEmpty-1201" src="https://user-images.githubusercontent.com/5115928/234139377-218ba2e4-0812-4810-84e3-a43806bec135.png">

#### Form Visible

<img width="1201" alt="ShoppingListsForm-1201" src="https://user-images.githubusercontent.com/5115928/234139398-97067227-cd74-497d-8ecb-350a67116717.png">

### 1405px

#### Form Hidden

<img width="1405" alt="ShoppingListsEmpty-1405" src="https://user-images.githubusercontent.com/5115928/234139424-7a1fcb32-de86-4559-b32c-2889cafde47e.png">

#### Form Visible

<img width="1405" alt="ShoppingListsForm-1405" src="https://user-images.githubusercontent.com/5115928/234139444-bef21c01-7b78-4c2a-a914-cd358d849e59.png">

### Large Desktop

#### Form Hidden

<img width="1919" alt="ShoppingListsEmpty-Full" src="https://user-images.githubusercontent.com/5115928/234139479-bd37e212-947c-4667-bc7e-eaac12f79a8c.png">

#### Form Visible

<img width="1919" alt="ShoppingListsForm-Full" src="https://user-images.githubusercontent.com/5115928/234139492-407b99f0-2adc-4b9e-9eea-c7e455894e1c.png">
